File: Serverko/scriptpool/scripts/framework/serverkobridge.js
console = 
{ 
	msgs: new Array(), log: function(msg) 
	{ print(msg); }, 
	info: function(msg) { print(msg); }, 
	warn : function(msg) { print(msg); }, 
	error: function(msg) { print(msg); }, 
	shift: function() { return console.msgs.shift(); } 
};
	

function serverko_startData()
{
    serverkob.startData();
    
}

function serverko_appendEvent(id, area, data)
{
    serverkob.appendEvent(id, area, data);
}

function serverko_sendData()
{
    serverkob.sendData();
}



function serverko_clientStartData(userID)
{
    serverkob.clientStartData(userID);
}

function serverko_clientAppendEvent(userID , id, area, data)
{
    serverkob.clientAppendEvent(userID , id, area, data)
}

function serverko_clientSendData(userID)
{
    serverkob.clientSendData(userID);
}

function serverko_disconnectClient(userID)
{
    serverkob.disconnectClient(userID);
}


function serverko_startTag(id) 
{
	serverkob.startTag(id);
}

function serverko_appendTag( id, data)
{
	serverkob._appendTag( id, data);
}

function serverko_endTag(id)
{
	serverkob.endTag(id);
}

function serverko_startTag2(id) 
{
    serverkob.startTag(id);

}

function serverko_appendTag2( id, data)
{
    serverkob.appendTag( id, data);
}

function serverko_startTags2( id, data)
{
    serverkob.startTags( id, data);
}

function serverko_endTags2(id)
{
    serverkob.endTags(id);
}

function serverko_appendSeparator2()
{
    serverkob.addSeparator();
}


function serverko_endTag2(id)
{
    serverkob.endTag(id);
}



function serverko_clientStartTag(userID , id) 
{
	serverkob.clientStartTag(userID , id);
}

function serverko_clientAppendTag( userID , id, data)
{
	serverkob.clientAppendTag( userID , id, data);
}

function serverko_clientEndTag(userID , id)
{
	serverkob.clientEndTag(userID , id);
}

function serverko_clientStartTag2(userID , id) 
{
    serverkob.clientStartTag(userID , id);

}

function serverko_clientAppendTag2( userID , id, data)
{
    serverkob.clientAppendTag( userID , id, data);
}

function serverko_clientStartTags2( userID , id)
{
    serverkob.clientStartTags( userID , id);
}

function serverko_clientEndTags2(userID , id)
{
    serverkob.clientEndTags(userID , id);
}

function serverko_clientAppendSeparator2(userID )
{
    serverkob.clientAddSeparator(userID );
}


function serverko_clientEndTag2(userID , id)
{
    serverkob.clientEndTag(userID , id);
}



function serverko_storeLayout(id)
{
	
}


function serverko_loadModule(id)
{
    serverkob.loadModule(id);
    
}
function serverko_loadModule2(id)
{
	serverkob.loadModule2(id);
    
}

function serverko_trace(str)
{
}

function serverko_clientSpectate(userID , yesno)
{
}

function serverko_sendEvent(id, delay, data) {
    serverkob.sendEvent(id, delay, data);
}

function serverko_exec(delay, script) {
    serverkob.exec(delay, script);
}

function serverko_exec_(delay, script) {
    //print("INFO: exec " + delay + " " + script);
    serverkob.exec(delay, script);
}

function Serverko()
{
	this.startData = serverko_startData;
	this.appendEvent = serverko_appendEvent;
	this.sendData =  serverko_sendData;
	this.clientStartData = serverko_clientStartData;
	this.clientAppendEvent =  serverko_clientAppendEvent;
	this.clientSendData =  serverko_clientSendData;
	this.disconnectClient =  serverko_disconnectClient;
	this.startTag =  serverko_startTag2;
	this.startTags =  serverko_startTags2;        
	this.appendTag = serverko_appendTag2;
	this.endTag = serverko_endTag2;
	this.endTags = serverko_endTags2;        
	this.addSeparator = serverko_appendSeparator2;
	
	this.clientStartTag =  serverko_clientStartTag2;
	this.clientStartTags =  serverko_clientStartTags2;        
	this.clientAppendTag = serverko_clientAppendTag2;
	this.clientEndTag = serverko_clientEndTag2;
	this.clientEndTags = serverko_clientEndTags2;        
	this.clientAddSeparator = serverko_clientAppendSeparator2;
	
			
	this.storeLayout = serverko_storeLayout;
	//this.getLayout = serverko_getLayout;
	this.loadModule = serverko_loadModule;
	this.loadModule2 = serverko_loadModule2;
    this.trace = serverko_trace;
    this.sendEvent = serverko_sendEvent;
    this.clientSpectate = serverko_clientSpectate;
    this.exec = serverko_exec;
    this.exec_ = serverko_exec_;

}

var serverko = new Serverko();

File: Serverko/scriptpool/scripts/framework/serverkobridge.py

#
#   Copyright 2012, Telum Slavonski Brod, Croatia.
#
#   Licensed under the Apache License, Version 2.0 (the "License")
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#   
#   This file is part of QFramework project, and can be used only as part of project.
#   Should be used for peace, not war :)   
#


class Console:
    def log(self,strdata):
        serverkob.trace("JS:"+strdata)
    def error(self,strdata):
        serverkob.trace("ERROR:"+strdata)
    def info(self,strdata):
        serverkob.trace("INFO:"+strdata)
    def warn(self,strdata):
        serverkob.trace("WARN:"+strdata)

console = Console()

console.log( "serverkobridgepy start ")

class Serverko:
    databuffer = []
    clientbuffer = {}

    def reserveSpace(self):
        if len(self.databuffer) == 0:
            self.databuffer.append( [] )
        arr = self.databuffer[-1]
        arr.append("")


    def startData(self):
        if ( len(self.databuffer) >  300):
            console.log("ERROR: data push large " + self.databuffer[3])
        self.databuffer.append( [])

    def appendEvent(self,id, area, data  = None, data2 = None, data3 = None, data4 = None, data5 = None):
        if len(self.databuffer) == 0:
            self.databuffer.append("")

        strdata = "";
            
        strdata += "{\"res\":\"event\",\"id\":\""+str(id)+"\",\"type\":\""+area;
            
        if data != None:
            strdata += "\",\"data\":\""+data
        if data2 != None:
            strdata += "\",\"data2\":\""+data2
        if data3 != None:
            strdata += "\",\"data3\":\""+data3
        if data4 != None:
            strdata += "\",\"data4\":\""+data4
            
        strdata += "\"}"
            
        self.databuffer[ -1 ].append( strdata );
    
    def appendEvent_(self , id, area, data  = None, data2 = None, data3 = None, data4 = None, data5 = None):
        self.startData()
        self.appendEvent(id , area, data , data2, data3, data4, data5)
        self.sendData()

    
    def clientStartData(self,userID):
        databuffer = self.clientbuffer[userID];
        if ( len(databuffer) >  300):
            console.log("ERROR: data push large " + databuffer[3])
        databuffer.append( [])


    def clientAppendEvent_(self , userID , id, area, data  = None, data2 = None, data3 = None, data4 = None, data5 = None):
        self.clientStartData(userID)
        self.clientAppendEvent(userID , id , area, data , data2, data3, data4, data5)
        self.clientSendData(userID)


    def clientAppendEvent(self,userID , id, area, data  = None, data2 = None, data3 = None, data4 = None, data5 = None):
        databuffer = self.clientbuffer[userID];
        if len(databuffer) == 0:
            databuffer.append("")

        strdata = "";
            
        strdata += "{\"res\":\"event\",\"id\":\""+str(id)+"\",\"type\":\""+area
            
        if data != None:
            strdata += "\",\"data\":\""+data
        if data2 != None:
            strdata += "\",\"data2\":\""+data2
        if data3 != None:
            strdata += "\",\"data3\":\""+data3
        if data4 != None:
            strdata += "\",\"data4\":\""+data4
        strdata += "\"}"
            
        databuffer[ -1 ].append( strdata );


    def startTag(self,id = None):
        strdata = "" 
        if (id != None):
            strdata = "{\""+id+"\": "
        else:
            strdata = "{";
        
        arr = self.databuffer[ -1 ] 
        arr[ -1 ] += strdata
    
    def startTags(self ,  id):
        arr = self.databuffer[  -1 ]
        arr[-1] += "\""+id+"\": ["; 
    
    def appendTag( self , id, data):
        arr = self.databuffer[ -1 ]
        arr[-1] += "\""+id+"\": " + "\""+data+"\""
    
    def endTag(self,id):
        arr = self.databuffer[ -1 ]
        arr[-1] += "}";
    
    def endTags(self,id):
        arr = self.databuffer[ -1 ]
        arr[-1] += "]"

    def addSeparator(self):
        arr = self.databuffer[ -1 ]
        arr[-1] += ","

    def toUser(self,userid):
        arr = self.databuffer[ -1 ]
        strdata = arr.pop(-1)
        if len(strdata) > 0:
            databuffer = self.clientbuffer[userid];
            databuffer[-1].append(strdata)
        return self
    
    def toUserNow(self):
        arr = self.databuffer[  -1 ]
        data = arr.pop(-1)
        if len(data) > 0:
            strdata = "{\"gs\":{\"room\":[" + data + "]}}\r\n" 
            self.clientSendData(strdata);
        return self

    def now(self):
        arr = self.databuffer[ -1 ]
        if len(arr) > 0:
            data = arr.pop(-1)
            if len(data) > 0:
                strdata = "{\"gs\":{\"room\":[" + data + "]}}\r\n" 
                serverkob.sendData(strdata)
        return self

    def evals(self,delay, script):
        self.lastData = ""
        serverkob.evals(delay, script)

    def evals_(self,delay, script):
        self.lastData = ""
        serverkob.evals(delay, script)

    def sendData(self):
        if len(self.databuffer) > 0:
            arr = self.databuffer.pop(-1)
            data = "" 
            while len(arr):
                data += arr.pop(0)
                if len(arr) > 0:
                    data += ","
            strdata = "{\"gs\":{\"room\":[" + data + "]}}\r\n" 
            serverkob.sendData(strdata)

        

    def clientSendData(self,userID):
        databuffer = self.clientbuffer[userID];
        if len(databuffer) > 0:
            arr = databuffer.pop(-1)
            data = "" 
            while len(arr):
                data += arr.pop(0);
                if len(arr) > 0:
                    data += ","
            strdata = "{\"gs\":{\"room\":[" + data + "]}}\r\n" 
            serverkob.clientSendData(userID , strdata)


    def sendEvent(selfid, delay, data):
        self.lastData = ""
        serverkob.sendEvent(id, delay, data)

    def disconnectClient(self,id):
        serverkob.disconnectClient(id)
    
    def loadModule(self,id):
        self.lastData = ""
        serverkob.loadModule(id)

    def loadModule2(self,id):
        self.lastData = ""
        serverkob.loadModule2(id)

    def onClientConnected(self,userID):
        self.clientbuffer[userID] = []

    def onClientDisconnected(self , userID):
        del self.clientbuffer[userID]

console.log( "serverkobridgepy OK ")
